File: python/Arrays&Hashing/group_anagrams.py
from typing import List
from collections import defaultdict

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        """
        Approach:
        - For each string, count the frequency of each character (assuming lowercase English letters).
        - Use the tuple of counts as a key in a dictionary to group anagrams together.
        - Return the grouped anagrams as a list of lists.

        Time Complexity: O(N * K), where N is the number of strings and K is the maximum length of a string.
        Space Complexity: O(N * K), for the result and the hash map.
        """
        res = defaultdict(list)
        for s in strs:
            count = [0] * 26
            for c in s:
                count[ord(c) - ord("a")] += 1 # ord() returns the ASCII value of the character.
            res[tuple(count)].append(s) # Use tuple as key since lists are not hashable.
        return list(res.values())

# -------------------
# Test Cases
# -------------------
def test_group_anagrams():
    sol = Solution()
    # Test case 1: Basic anagrams
    strs1 = ["eat", "tea", "tan", "ate", "nat", "bat"]
    result1 = sol.groupAnagrams(strs1)
    assert sorted([sorted(group) for group in result1]) == sorted([sorted(group) for group in [["eat","tea","ate"],["tan","nat"],["bat"]]])
    # Test case 2: Single string
    strs2 = ["abc"]
    assert sol.groupAnagrams(strs2) == [["abc"]]
    # Test case 3: Empty list
    strs3 = []
    assert sol.groupAnagrams(strs3) == []
    # Test case 4: All strings are anagrams
    strs4 = ["abc", "bca", "cab"]
    result4 = sol.groupAnagrams(strs4)
    assert sorted([sorted(group) for group in result4]) == [["abc", "bca", "cab"]]
    # Test case 5: No anagrams
    strs5 = ["a", "b", "c"]
    result5 = sol.groupAnagrams(strs5)
    assert sorted([sorted(group) for group in result5]) == [["a"], ["b"], ["c"]]
    print("All test cases passed!")

if __name__ == "__main__":
    test_group_anagrams()
    

File: C++/Arrays&Hashing/group_anagrams.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    /*
     * Problem: Group Anagrams
     * 
     * Approach:
     * - Create unique key for each anagram group using character counts
     * - Use hash map to group strings with same character frequency
     * - For each string:
     *   1. Count frequency of each character (a-z)
     *   2. Create key from count array
     *   3. Add string to corresponding group
     * 
     * Time Complexity: O(n * k)
     * - n = number of strings
     * - k = maximum length of a string
     * - For each string, we count k characters
     * - Creating key takes O(1) as count array is fixed size (26)
     * 
     * Space Complexity: O(n * k)
     * - Storing all strings in hash map
     * - Each string appears exactly once
     */
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> groups;
        
        for (const string& s : strs) {
            vector<int> count(26, 0);
            for (char c : s) {
                count[c - 'a']++;
            }
            
            string key;
            for (int n : count) {
                key += to_string(n) + "#";
            }
            
            groups[key].push_back(s);
        }
        
        vector<vector<string>> result;
        for (const auto& pair : groups) {
            result.push_back(pair.second);
        }
        
        return result;
    }
};

/*
 * Detailed Syntax Explanation:
 * 
 * 1. Function Declaration:
 *    vector<vector<string>> groupAnagrams(vector<string>& strs)
 *    - Returns nested vector of strings (groups of anagrams)
 *    - Takes reference to vector of strings as input
 *    - & prevents copying of input vector
 * 
 * 2. HashMap Declaration:
 *    unordered_map<string, vector<string>> groups
 *    - Key: string (character frequency signature)
 *    - Value: vector of strings (anagram group)
 *    - O(1) average lookup/insert
 * 
 * 3. Character Counting:
 *    vector<int> count(26, 0)
 *    - Fixed size array for lowercase letters
 *    - Initialized with zeros
 *    count[c - 'a']++
 *    - Maps 'a'-'z' to indices 0-25
 * 
 * 4. Key Generation:
 *    string key
 *    key += to_string(n) + "#"
 *    - Converts count array to string
 *    - Uses "#" as delimiter between counts
 * 
 * 5. Group Storage:
 *    groups[key].push_back(s)
 *    - Adds string to its anagram group
 *    - Creates new vector if key doesn't exist
 * 
 * 6. Result Collection:
 *    for (const auto& pair : groups)
 *    - Iterates through all groups
 *    - pair.first is key (unused)
 *    - pair.second is vector of anagrams
 */